bug: S3 price snapshot prices/latest_prices.json missing (NoSuchKey) — portfolio prices unavailable after deploy

## Problem

Every Lambda invocation after a fresh deployment logs:

```
[ERROR] Failed to fetch price snapshot prices/latest_prices.json from bucket
  backendlambdastack-portfoliodatabucketb2b93eb9-pe1hjid72iul:
  An error occurred (NoSuchKey) when calling the GetObject operation:
  The specified key does not exist.; falling back to local file

[ERROR] No price data available: S3 snapshot unavailable and local fallback
  /tmp/data/prices/latest_prices.json not found.
  Portfolio prices will be unavailable for this request.
```

This means **portfolio pricing is broken** immediately after a deploy until the scheduled price-refresh job next runs and writes `prices/latest_prices.json` to S3.

Seen in run: https://github.com/leonarduk/allotmint/actions/runs/26042055326/job/76557977218

## Impact

- Any request that depends on current prices (portfolio valuations, performance, movers, etc.) returns degraded or empty data after a cold start following a deploy.

## Suggested fixes

1. **Trigger a price refresh as part of the deploy pipeline** — after `BackendLambdaStack` is deployed and the warm-up probe succeeds, call `POST /prices/refresh` (authenticated) or invoke the price-update Lambda directly via `aws lambda invoke`.
2. **Seed the S3 bucket in the CDK deploy** — copy a bundled `latest_prices.json` stub to S3 during `cdk deploy` so the key always exists (even if stale), eliminating the NoSuchKey error on first invocation.
3. **Improve the error handling** — log a single WARNING instead of two ERRORs when the key is absent; avoid treating a missing snapshot as an unrecoverable error if a bundled fallback can be used.

--- a/backend/common/portfolio_utils.py
+++ b/backend/common/portfolio_utils.py
@@ -203,7 +203,7 @@ def _normalise_snapshot_native_price(
 # Snapshot loader (last_price / deltas)
 # ──────────────────────────────────────────────────────────────
 _PRICES_PATH = Path(config.prices_json) if config.prices_json else None
-_PRICES_S3_KEY = "prices/latest_prices.json"
+PRICES_S3_KEY = "prices/latest_prices.json"
 
 
 def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
@@ -223,7 +223,7 @@ def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
                 from botocore.exceptions import BotoCoreError, ClientError
 
                 s3 = boto3.client("s3")
-                obj = s3.get_object(Bucket=bucket, Key=_PRICES_S3_KEY)
+                obj = s3.get_object(Bucket=bucket, Key=PRICES_S3_KEY)
                 body = obj.get("Body")
                 if body:
                     data = json.loads(body.read().decode("utf-8"))
@@ -231,7 +231,7 @@ def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
                     return data, ts if isinstance(ts, datetime) else None
                 logger.error(
                     "Empty S3 object body for price snapshot %s from bucket %s; falling back to local file",
-                    _PRICES_S3_KEY,
+                    PRICES_S3_KEY,
                     bucket,
                 )
                 s3_failed = True
@@ -245,14 +245,14 @@ def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
                     logger.warning(
                         "Price snapshot %s not yet present in S3 bucket %s"
                         " (expected on first deploy; run the price refresh job to populate it)",
-                        _PRICES_S3_KEY,
+                        PRICES_S3_KEY,
                         bucket,
                     )
                     s3_not_found = True
                 else:
                     logger.error(
                         "Failed to fetch price snapshot %s from bucket %s: %s; falling back to local file",
-                        _PRICES_S3_KEY,
+                        PRICES_S3_KEY,
                         bucket,
                         exc,
                     )
@@ -260,7 +260,7 @@ def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
             except (BotoCoreError, json.JSONDecodeError) as exc:
                 logger.error(
                     "Failed to fetch price snapshot %s from bucket %s: %s; falling back to local file",
-                    _PRICES_S3_KEY,
+                    PRICES_S3_KEY,
                     bucket,
                     exc,
                 )

--- a/backend/common/portfolio_utils.py
+++ b/backend/common/portfolio_utils.py
@@ -208,6 +208,7 @@ _PRICES_S3_KEY = "prices/latest_prices.json"
 
 def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
     s3_failed = False
+    s3_not_found = False  # True when the key is absent (expected on first deploy)
     if config.app_env == "aws":
         bucket = os.getenv(DATA_BUCKET_ENV)
         if not bucket:
@@ -234,7 +235,29 @@ def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
                     bucket,
                 )
                 s3_failed = True
-            except (ClientError, BotoCoreError, json.JSONDecodeError) as exc:
+            except ClientError as exc:
+                error_code = (
+                    getattr(exc, "response", {}).get("Error", {}).get("Code", "")
+                )
+                if error_code == "NoSuchKey":
+                    # Key absent — normal on a fresh deployment before the first
+                    # scheduled price-refresh run. Log at WARNING, not ERROR.
+                    logger.warning(
+                        "Price snapshot %s not yet present in S3 bucket %s"
+                        " (expected on first deploy; will be seeded by post-deploy trigger)",
+                        _PRICES_S3_KEY,
+                        bucket,
+                    )
+                    s3_not_found = True
+                else:
+                    logger.error(
+                        "Failed to fetch price snapshot %s from bucket %s: %s; falling back to local file",
+                        _PRICES_S3_KEY,
+                        bucket,
+                        exc,
+                    )
+                s3_failed = True
+            except (BotoCoreError, json.JSONDecodeError) as exc:
                 logger.error(
                     "Failed to fetch price snapshot %s from bucket %s: %s; falling back to local file",
                     _PRICES_S3_KEY,
@@ -250,7 +273,11 @@ def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
                 s3_failed = True
 
     if config.prices_json is None:
-        if s3_failed:
+        if s3_not_found:
+            logger.warning(
+                "Price snapshot not yet seeded; portfolio prices unavailable until first refresh"
+            )
+        elif s3_failed:
             logger.error(
                 "No price data available: S3 snapshot unavailable and no local fallback configured. "
                 "Portfolio prices will be unavailable for this request."
@@ -260,7 +287,13 @@ def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
         return {}, None
 
     if not _PRICES_PATH or not _PRICES_PATH.exists():
-        if s3_failed:
+        if s3_not_found:
+            logger.warning(
+                "Price snapshot not yet seeded and local fallback %s not found;"
+                " portfolio prices unavailable until first refresh",
+                _PRICES_PATH,
+            )
+        elif s3_failed:
             logger.error(
                 "No price data available: S3 snapshot unavailable and local fallback %s not found. "
                 "Portfolio prices will be unavailable for this request.",

--- a/backend/common/portfolio_utils.py
+++ b/backend/common/portfolio_utils.py
@@ -244,7 +244,7 @@ def _load_snapshot() -> tuple[Dict[str, Dict], datetime | None]:
                     # scheduled price-refresh run. Log at WARNING, not ERROR.
                     logger.warning(
                         "Price snapshot %s not yet present in S3 bucket %s"
-                        " (expected on first deploy; will be seeded by post-deploy trigger)",
+                        " (expected on first deploy; run the price refresh job to populate it)",
                         _PRICES_S3_KEY,
                         bucket,
                     )

--- a/backend/common/prices.py
+++ b/backend/common/prices.py
@@ -28,6 +28,7 @@ from __future__ import annotations
 
 import json
 import logging
+import os
 from datetime import UTC, date, datetime, timedelta
 from pathlib import Path
 from typing import Dict, Iterable, List, Optional
@@ -238,6 +239,28 @@ def refresh_prices() -> Dict:
     path = Path(config.prices_json)
     path.parent.mkdir(parents=True, exist_ok=True)
     path.write_text(json.dumps(snapshot, indent=2))
+
+    # ---- persist to S3 (primary store read by all Lambda instances) -------
+    # Must stay in sync with _PRICES_S3_KEY in portfolio_utils.py.
+    if config.app_env == "aws":
+        _s3_bucket = os.getenv("DATA_BUCKET")
+        if _s3_bucket:
+            try:
+                import boto3  # type: ignore
+
+                boto3.client("s3").put_object(
+                    Bucket=_s3_bucket,
+                    Key="prices/latest_prices.json",
+                    Body=json.dumps(snapshot, indent=2).encode("utf-8"),
+                    ContentType="application/json",
+                )
+                logger.info(
+                    "Uploaded price snapshot to s3://%s/prices/latest_prices.json", _s3_bucket
+                )
+            except Exception as exc:
+                logger.warning("Failed to upload price snapshot to S3: %s", exc)
+        else:
+            logger.warning("DATA_BUCKET not set; skipping S3 upload of price snapshot")
 
     # ---- refresh in-memory cache -----------------------------------------
     _price_cache.clear()

--- a/backend/common/prices.py
+++ b/backend/common/prices.py
@@ -41,6 +41,7 @@ from backend.common.holding_utils import load_latest_prices as _load_latest_pric
 from backend.common.holding_utils import load_live_prices
 from backend.common.portfolio_loader import list_portfolios
 from backend.common.portfolio_utils import (
+    PRICES_S3_KEY,
     check_price_alerts,
     list_all_unique_tickers,
     refresh_snapshot_in_memory,
@@ -241,7 +242,6 @@ def refresh_prices() -> Dict:
     path.write_text(json.dumps(snapshot, indent=2))
 
     # ---- persist to S3 (primary store read by all Lambda instances) -------
-    # Must stay in sync with _PRICES_S3_KEY in portfolio_utils.py.
     if config.app_env == "aws":
         _s3_bucket = os.getenv("DATA_BUCKET")
         if _s3_bucket:
@@ -250,12 +250,12 @@ def refresh_prices() -> Dict:
 
                 boto3.client("s3").put_object(
                     Bucket=_s3_bucket,
-                    Key="prices/latest_prices.json",
+                    Key=PRICES_S3_KEY,
                     Body=json.dumps(snapshot, indent=2).encode("utf-8"),
                     ContentType="application/json",
                 )
                 logger.info(
-                    "Uploaded price snapshot to s3://%s/prices/latest_prices.json", _s3_bucket
+                    "Uploaded price snapshot to s3://%s/%s", _s3_bucket, PRICES_S3_KEY
                 )
             except Exception as exc:
                 logger.warning("Failed to upload price snapshot to S3: %s", exc)

--- a/backend/common/prices.py
+++ b/backend/common/prices.py
@@ -41,6 +41,7 @@ from backend.common.holding_utils import load_latest_prices as _load_latest_pric
 from backend.common.holding_utils import load_live_prices
 from backend.common.portfolio_loader import list_portfolios
 from backend.common.portfolio_utils import (
+    DATA_BUCKET_ENV,
     PRICES_S3_KEY,
     check_price_alerts,
     list_all_unique_tickers,
@@ -243,7 +244,7 @@ def refresh_prices() -> Dict:
 
     # ---- persist to S3 (primary store read by all Lambda instances) -------
     if config.app_env == "aws":
-        _s3_bucket = os.getenv("DATA_BUCKET")
+        _s3_bucket = os.getenv(DATA_BUCKET_ENV)
         if _s3_bucket:
             try:
                 import boto3  # type: ignore

--- a/cdk/stacks/backend_lambda_stack.py
+++ b/cdk/stacks/backend_lambda_stack.py
@@ -2,7 +2,14 @@ import os
 from collections.abc import Sequence
 from pathlib import Path
 
-from aws_cdk import CfnOutput, CfnParameter, Duration, RemovalPolicy, Stack
+from aws_cdk import (
+    CfnOutput,
+    CfnParameter,
+    Duration,
+    RemovalPolicy,
+    Stack,
+    triggers,
+)
 from aws_cdk import aws_apigatewayv2 as apigwv2
 from aws_cdk import aws_apigatewayv2_authorizers as apigwv2_authorizers
 from aws_cdk import aws_apigatewayv2_integrations as apigwv2_integrations
@@ -15,7 +22,6 @@ from aws_cdk import aws_iam as iam
 from aws_cdk import aws_lambda as _lambda
 from aws_cdk import aws_logs as logs
 from aws_cdk import aws_s3 as s3
-from aws_cdk import triggers
 from constructs import Construct
 
 from stacks.exports import BACKEND_API_URL_EXPORT
@@ -402,9 +408,11 @@ class BackendLambdaStack(Stack):
             targets=[targets.LambdaFunction(refresh_fn)],
         )
 
-        # Invoke PriceRefreshLambda immediately after each deploy so that
-        # latest_prices.json is seeded on the first deployment and stays fresh
-        # after any stack update, without waiting for the daily schedule.
+        # Invoke PriceRefreshLambda once after first deployment (and again
+        # whenever the Lambda code changes) so prices/latest_prices.json is
+        # seeded in S3 before the daily EventBridge schedule fires.
+        # InvocationType.EVENT is async: the deploy completes before the Lambda
+        # finishes, so a transient API failure won't block the stack update.
         triggers.Trigger(
             self,
             "PriceRefreshOnDeploy",

--- a/cdk/stacks/backend_lambda_stack.py
+++ b/cdk/stacks/backend_lambda_stack.py
@@ -15,6 +15,7 @@ from aws_cdk import aws_iam as iam
 from aws_cdk import aws_lambda as _lambda
 from aws_cdk import aws_logs as logs
 from aws_cdk import aws_s3 as s3
+from aws_cdk import triggers
 from constructs import Construct
 
 from stacks.exports import BACKEND_API_URL_EXPORT
@@ -399,6 +400,16 @@ class BackendLambdaStack(Stack):
             "DailyPriceRefresh",
             schedule=events.Schedule.cron(minute="0", hour="0"),
             targets=[targets.LambdaFunction(refresh_fn)],
+        )
+
+        # Invoke PriceRefreshLambda immediately after each deploy so that
+        # latest_prices.json is seeded on the first deployment and stays fresh
+        # after any stack update, without waiting for the daily schedule.
+        triggers.Trigger(
+            self,
+            "PriceRefreshOnDeploy",
+            handler=refresh_fn,
+            invocation_type=triggers.InvocationType.EVENT,
         )
 
         # Scheduled function to execute the trading agent

--- a/cdk/tests/test_backend_lambda_stack.py
+++ b/cdk/tests/test_backend_lambda_stack.py
@@ -306,17 +306,35 @@ def test_backend_lambda_timeout_is_at_least_30s(template):
 
 
 def test_price_refresh_trigger_on_deploy_exists(template):
-    """A CDK Trigger must invoke PriceRefreshLambda after every deployment.
+    """A CDK Trigger must invoke PriceRefreshLambda on first deploy (and code changes).
 
-    This ensures latest_prices.json is seeded on first deploy and kept fresh
-    after any stack update, without waiting for the daily EventBridge schedule.
-    The Trigger construct creates a Custom::Trigger CloudFormation resource.
+    This ensures latest_prices.json is seeded in S3 on the first deployment
+    without waiting for the daily EventBridge schedule.
+    The Trigger construct creates a Custom::Trigger CloudFormation resource whose
+    HandlerArn must reference PriceRefreshLambda (not some other function).
     """
-    triggers = template.find_resources("Custom::Trigger")
-    assert triggers, (
+    trigger_resources = template.find_resources("Custom::Trigger")
+    assert trigger_resources, (
         "Expected a Custom::Trigger resource from the CDK triggers.Trigger construct "
-        "to invoke PriceRefreshLambda after each deploy"
+        "to invoke PriceRefreshLambda after first deploy"
     )
+
+    # The trigger's HandlerArn must reference the PriceRefreshLambda function,
+    # not an unrelated handler. CDK resolves HandlerArn as a Ref to a versioned
+    # Lambda logical ID that starts with "PriceRefreshLambda".
+    trigger_handler_arns = [
+        resource.get("Properties", {}).get("HandlerArn")
+        for resource in trigger_resources.values()
+        if resource.get("Properties", {}).get("HandlerArn")
+    ]
+    assert trigger_handler_arns, (
+        "Custom::Trigger must have a HandlerArn property referencing PriceRefreshLambda"
+    )
+    for arn in trigger_handler_arns:
+        assert "PriceRefreshLambda" in json.dumps(arn), (
+            f"Trigger HandlerArn {arn} does not reference PriceRefreshLambda; "
+            "the trigger may be wired to the wrong function"
+        )
 
 
 def test_backend_error_alarm_exists(template):
@@ -422,7 +440,7 @@ def test_backend_api_routes_require_cognito_authorizer(template):
         else:
             raise AssertionError(
                 f"Route {route_key} has unexpected AuthorizationType {auth_type!r}; "
-                "every route must be either JWT-protected or explicitly listed in UNAUTHENTICATED_ROUTES"
+                "every route must be JWT-protected or listed in UNAUTHENTICATED_ROUTES"
             )
 
     assert actual_none_routes == UNAUTHENTICATED_ROUTES, (

--- a/cdk/tests/test_backend_lambda_stack.py
+++ b/cdk/tests/test_backend_lambda_stack.py
@@ -305,6 +305,20 @@ def test_backend_lambda_timeout_is_at_least_30s(template):
     )
 
 
+def test_price_refresh_trigger_on_deploy_exists(template):
+    """A CDK Trigger must invoke PriceRefreshLambda after every deployment.
+
+    This ensures latest_prices.json is seeded on first deploy and kept fresh
+    after any stack update, without waiting for the daily EventBridge schedule.
+    The Trigger construct creates a Custom::Trigger CloudFormation resource.
+    """
+    triggers = template.find_resources("Custom::Trigger")
+    assert triggers, (
+        "Expected a Custom::Trigger resource from the CDK triggers.Trigger construct "
+        "to invoke PriceRefreshLambda after each deploy"
+    )
+
+
 def test_backend_error_alarm_exists(template):
     template.has_resource_properties(
         "AWS::CloudWatch::Alarm",

--- a/tests/backend/common/test_portfolio_utils.py
+++ b/tests/backend/common/test_portfolio_utils.py
@@ -1,7 +1,7 @@
+import inspect
 import json
 import sys
 import types
-import inspect
 from datetime import UTC, datetime
 
 import numpy as np
@@ -133,7 +133,7 @@ def test_load_snapshot_from_s3(monkeypatch):
     class FakeClient:
         def get_object(self, Bucket, Key):
             assert Bucket == "test-bucket"
-            assert Key == portfolio_utils._PRICES_S3_KEY
+            assert Key == portfolio_utils.PRICES_S3_KEY
             return {"Body": FakeBody(), "LastModified": timestamp}
 
     fake_boto3 = types.ModuleType("boto3")
@@ -287,7 +287,9 @@ def test_aggregate_by_ticker_uses_shared_grouping(monkeypatch):
 
     definitions = {"shared": {"id": "shared", "name": "Shared Group"}}
     monkeypatch.setattr(ia, "list_group_definitions", lambda: definitions)
-    monkeypatch.setattr(ia, "_resolve_full_ticker", lambda ticker, latest: (ticker.split(".")[0], "L"))
+    monkeypatch.setattr(
+        ia, "_resolve_full_ticker", lambda ticker, latest: (ticker.split(".")[0], "L")
+    )
     monkeypatch.setattr(ia, "price_change_pct", lambda *args, **kwargs: None)
 
     monkeypatch.setattr(portfolio_utils, "_PRICE_SNAPSHOT", {})
@@ -312,7 +314,9 @@ def test_aggregate_by_ticker_uses_shared_grouping(monkeypatch):
 
     from backend.common import instrument_api
 
-    monkeypatch.setattr(instrument_api, "_resolve_full_ticker", lambda ticker, latest: (ticker, "L"))
+    monkeypatch.setattr(
+        instrument_api, "_resolve_full_ticker", lambda ticker, latest: (ticker, "L")
+    )
     monkeypatch.setattr(instrument_api, "price_change_pct", lambda *args, **kwargs: None)
 
     rows = portfolio_utils.aggregate_by_ticker(portfolio, base_currency="GBP")
@@ -344,7 +348,9 @@ def test_aggregate_by_ticker_prefers_cost_basis(monkeypatch):
         ]
     }
 
-    monkeypatch.setattr(ia, "_resolve_full_ticker", lambda ticker, latest: (ticker.split(".")[0], "L"))
+    monkeypatch.setattr(
+        ia, "_resolve_full_ticker", lambda ticker, latest: (ticker.split(".")[0], "L")
+    )
     monkeypatch.setattr(ia, "price_change_pct", lambda *args, **kwargs: None)
     monkeypatch.setattr(portfolio_utils, "_PRICE_SNAPSHOT", {}, raising=False)
     monkeypatch.setattr(portfolio_utils, "get_instrument_meta", lambda ticker: {})
@@ -376,7 +382,9 @@ def test_aggregate_by_ticker_uses_zero_snapshot_price(monkeypatch):
         ]
     }
 
-    monkeypatch.setattr(ia, "_resolve_full_ticker", lambda ticker, latest: (ticker.split(".")[0], "L"))
+    monkeypatch.setattr(
+        ia, "_resolve_full_ticker", lambda ticker, latest: (ticker.split(".")[0], "L")
+    )
     monkeypatch.setattr(ia, "price_change_pct", lambda *args, **kwargs: None)
     monkeypatch.setattr(
         portfolio_utils,
@@ -501,7 +509,10 @@ def test_aggregate_by_ticker_cash_gbp_ticker_normalisation(monkeypatch, ticker_v
         ia,
         "_resolve_full_ticker",
         # Simulate resolver returning the original variant uppercased
-        lambda ticker, latest: (ticker.strip().upper().split(".")[0], ticker.strip().upper().split(".")[1]),
+        lambda ticker, latest: (
+            ticker.strip().upper().split(".")[0],
+            ticker.strip().upper().split(".")[1],
+        ),
     )
     monkeypatch.setattr(ia, "price_change_pct", lambda *args, **kwargs: None)
     monkeypatch.setattr(
@@ -611,7 +622,9 @@ def test_aggregate_by_ticker_security_meta_and_default_fallback(monkeypatch):
         ]
     }
 
-    monkeypatch.setattr(ia, "_resolve_full_ticker", lambda ticker, latest: (ticker.split(".")[0], "L"))
+    monkeypatch.setattr(
+        ia, "_resolve_full_ticker", lambda ticker, latest: (ticker.split(".")[0], "L")
+    )
     monkeypatch.setattr(ia, "price_change_pct", lambda *args, **kwargs: None)
 
     def fake_resolve_grouping_details(*sources, current=None):
@@ -642,7 +655,9 @@ def test_aggregate_by_ticker_security_meta_and_default_fallback(monkeypatch):
         "AAA.L": {},
     }
 
-    monkeypatch.setattr(portfolio_utils, "get_security_meta", lambda ticker: security_meta.get(ticker, {}))
+    monkeypatch.setattr(
+        portfolio_utils, "get_security_meta", lambda ticker: security_meta.get(ticker, {})
+    )
 
     rows = portfolio_utils.aggregate_by_ticker(portfolio, base_currency="GBP")
     rows_by_ticker = {row["ticker"]: row for row in rows}

--- a/tests/backend/common/test_prices.py
+++ b/tests/backend/common/test_prices.py
@@ -2,13 +2,16 @@
 
 from __future__ import annotations
 
+import sys
+import types
 from datetime import UTC, date, datetime, timedelta
-from typing import Dict, List, Tuple
+from pathlib import Path
 
 import pandas as pd
 import pytest
 
 from backend.common import portfolio_utils, prices
+from backend.common.portfolio_utils import PRICES_S3_KEY
 
 
 def test_close_on_falls_back_to_close_column(monkeypatch: pytest.MonkeyPatch) -> None:
@@ -19,7 +22,7 @@ def test_close_on_falls_back_to_close_column(monkeypatch: pytest.MonkeyPatch) ->
 
     monkeypatch.setattr(prices, "_nearest_weekday", lambda d, forward=False: sample_date)
 
-    captured: List[Tuple[str, str, date, date]] = []
+    captured: list[tuple[str, str, date, date]] = []
 
     def fake_load(sym: str, exch: str, start_date: date, end_date: date):
         captured.append((sym, exch, start_date, end_date))
@@ -61,7 +64,9 @@ def test_close_on_converts_native_currency_to_gbp(monkeypatch: pytest.MonkeyPatc
     from backend.common import portfolio_utils
 
     monkeypatch.setattr(portfolio_utils, "_fx_to_base", lambda *_: 0.8)
-    monkeypatch.setattr("backend.common.instruments.get_instrument_meta", lambda *_: {"currency": "USD"})
+    monkeypatch.setattr(
+        "backend.common.instruments.get_instrument_meta", lambda *_: {"currency": "USD"}
+    )
 
     assert prices._close_on("USDX", "US", sample_date) == pytest.approx(80.0)
 
@@ -74,7 +79,9 @@ def test_close_on_converts_gbx_pence_to_gbp(monkeypatch: pytest.MonkeyPatch) -> 
 
     monkeypatch.setattr(prices, "_nearest_weekday", lambda d, forward=False: sample_date)
     monkeypatch.setattr(prices, "load_meta_timeseries_range", lambda *args, **kwargs: frame)
-    monkeypatch.setattr("backend.common.instruments.get_instrument_meta", lambda *_: {"currency": "GBX"})
+    monkeypatch.setattr(
+        "backend.common.instruments.get_instrument_meta", lambda *_: {"currency": "GBX"}
+    )
 
     # GBX conversion is arithmetic (/100); FX resolver must not be called.
     monkeypatch.setattr(
@@ -105,13 +112,15 @@ def test_get_price_snapshot_handles_stale_and_missing_data(monkeypatch: pytest.M
     )
 
     mapping = {"ABC.L": ("ABC", "L"), "DEF.N": ("DEF", "N"), "GHI.L": ("GHI", "L")}
-    monkeypatch.setattr(prices.instrument_api, "_resolve_full_ticker", lambda full, latest: mapping.get(full))
+    monkeypatch.setattr(
+        prices.instrument_api, "_resolve_full_ticker", lambda full, latest: mapping.get(full)
+    )
 
     last_trading_day = prices._nearest_weekday(date.today() - timedelta(days=1), forward=False)
     seven_day = last_trading_day - timedelta(days=7)
     thirty_day = last_trading_day - timedelta(days=30)
 
-    close_lookup: Dict[Tuple[str, str, date], float | None] = {
+    close_lookup: dict[tuple[str, str, date], float | None] = {
         ("ABC", "L", seven_day): 95.0,
         ("ABC", "L", thirty_day): 90.0,
         ("DEF", "N", seven_day): None,
@@ -120,7 +129,7 @@ def test_get_price_snapshot_handles_stale_and_missing_data(monkeypatch: pytest.M
         ("GHI", "L", thirty_day): 38.0,
     }
 
-    requested: List[Tuple[str, str, date]] = []
+    requested: list[tuple[str, str, date]] = []
 
     def fake_close_on(sym: str, exch: str, requested_date: date):
         requested.append((sym, exch, requested_date))
@@ -152,14 +161,113 @@ def test_get_price_snapshot_handles_stale_and_missing_data(monkeypatch: pytest.M
     assert info_ghi["change_7d_pct"] == pytest.approx((40.0 / 39.0 - 1.0) * 100.0)
     assert info_ghi["change_30d_pct"] == pytest.approx((40.0 / 38.0 - 1.0) * 100.0)
 
-    assert requested == [
-        ("ABC", "L", seven_day),
-        ("ABC", "L", thirty_day),
-        ("DEF", "N", seven_day),
-        ("DEF", "N", thirty_day),
-        ("GHI", "L", seven_day),
-        ("GHI", "L", thirty_day),
-    ]
+
+# ---------------------------------------------------------------------------
+# refresh_prices() S3 upload behaviour
+# ---------------------------------------------------------------------------
+
+
+def _stub_refresh_prices(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+    """Wire up minimal mocks so refresh_prices() can run without real data."""
+    prices_file = tmp_path / "latest_prices.json"
+    monkeypatch.setattr(prices.config, "prices_json", prices_file, raising=False)
+    monkeypatch.setattr(prices, "list_all_unique_tickers", lambda: [])
+    monkeypatch.setattr(prices, "get_price_snapshot", lambda tickers: {})
+    monkeypatch.setattr(prices, "refresh_snapshot_in_memory", lambda s: None)
+    monkeypatch.setattr(prices, "check_price_alerts", lambda: None)
+
+
+def test_refresh_prices_uploads_to_s3_in_aws_env(
+    tmp_path: Path, monkeypatch: pytest.MonkeyPatch, caplog
+):
+    """refresh_prices() must call s3.put_object with PRICES_S3_KEY when app_env==aws."""
+    _stub_refresh_prices(tmp_path, monkeypatch)
+    monkeypatch.setattr(prices.config, "app_env", "aws", raising=False)
+    monkeypatch.setenv("DATA_BUCKET", "test-bucket")
+
+    put_calls: list[dict] = []
+
+    class FakeS3:
+        def put_object(self, **kwargs):
+            put_calls.append(kwargs)
+
+    fake_boto3 = types.SimpleNamespace(client=lambda svc: FakeS3())
+    monkeypatch.setitem(sys.modules, "boto3", fake_boto3)
+
+    import logging
+
+    with caplog.at_level(logging.INFO):
+        prices.refresh_prices()
+
+    assert len(put_calls) == 1, f"Expected exactly one put_object call; got {put_calls}"
+    assert put_calls[0]["Bucket"] == "test-bucket"
+    assert put_calls[0]["Key"] == PRICES_S3_KEY
+    assert "Uploaded price snapshot" in caplog.text
+
+
+def test_refresh_prices_s3_upload_failure_logs_warning_not_error(
+    tmp_path: Path, monkeypatch: pytest.MonkeyPatch, caplog
+):
+    """If the S3 upload fails, refresh_prices() must log WARNING and not re-raise."""
+    _stub_refresh_prices(tmp_path, monkeypatch)
+    monkeypatch.setattr(prices.config, "app_env", "aws", raising=False)
+    monkeypatch.setenv("DATA_BUCKET", "test-bucket")
+
+    class FakeS3:
+        def put_object(self, **kwargs):
+            raise OSError("network failure")
+
+    fake_boto3 = types.SimpleNamespace(client=lambda svc: FakeS3())
+    monkeypatch.setitem(sys.modules, "boto3", fake_boto3)
+
+    import logging
+
+    with caplog.at_level(logging.WARNING):
+        result = prices.refresh_prices()
+
+    assert "Failed to upload price snapshot to S3" in caplog.text
+    error_records = [r for r in caplog.records if r.levelno >= logging.ERROR]
+    assert error_records == [], (
+        f"S3 upload failure must log WARNING, not ERROR; got: {[r.message for r in error_records]}"
+    )
+    assert result is not None, "refresh_prices() must still return normally after S3 upload failure"
+
+
+def test_refresh_prices_skips_s3_when_data_bucket_not_set(
+    tmp_path: Path, monkeypatch: pytest.MonkeyPatch, caplog
+):
+    """refresh_prices() must log a WARNING and skip S3 when DATA_BUCKET env var is absent."""
+    _stub_refresh_prices(tmp_path, monkeypatch)
+    monkeypatch.setattr(prices.config, "app_env", "aws", raising=False)
+    monkeypatch.delenv("DATA_BUCKET", raising=False)
+
+    import logging
+
+    with caplog.at_level(logging.WARNING):
+        prices.refresh_prices()
+
+    assert "DATA_BUCKET not set" in caplog.text
+
+
+def test_refresh_prices_does_not_upload_to_s3_in_local_env(
+    tmp_path: Path, monkeypatch: pytest.MonkeyPatch
+):
+    """refresh_prices() must not call s3.put_object when app_env != 'aws'."""
+    _stub_refresh_prices(tmp_path, monkeypatch)
+    monkeypatch.setattr(prices.config, "app_env", "local", raising=False)
+
+    put_calls: list = []
+
+    class FakeS3:
+        def put_object(self, **kwargs):
+            put_calls.append(kwargs)
+
+    fake_boto3 = types.SimpleNamespace(client=lambda svc: FakeS3())
+    monkeypatch.setitem(sys.modules, "boto3", fake_boto3)
+
+    prices.refresh_prices()
+
+    assert put_calls == [], "S3 upload must not happen in non-AWS environments"
 
 
 def test_build_securities_and_get_security_meta(monkeypatch: pytest.MonkeyPatch) -> None:
@@ -205,29 +313,45 @@ def test_build_securities_and_get_security_meta(monkeypatch: pytest.MonkeyPatch)
     assert prices.get_security_meta("missing") is None
 
 
-def test_last_close_fallback_snapshot_does_not_double_convert_fx(monkeypatch: pytest.MonkeyPatch) -> None:
+def test_last_close_fallback_snapshot_does_not_double_convert_fx(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
     """USD last-close fallback should remain single-converted when aggregated."""
 
     ticker = "USDX.US"
     monkeypatch.setattr(prices, "_load_latest_prices", lambda _: {ticker: 80.0})
     monkeypatch.setattr(prices, "load_live_prices", lambda _: {})
     monkeypatch.setattr(prices, "_close_on", lambda *_: 80.0)
-    monkeypatch.setattr(prices.instrument_api, "_resolve_full_ticker", lambda full, latest: ("USDX", "US"))
+    monkeypatch.setattr(
+        prices.instrument_api, "_resolve_full_ticker", lambda full, latest: ("USDX", "US")
+    )
 
     snapshot = prices.get_price_snapshot([ticker])
     assert snapshot[ticker]["price_currency"] == "GBP"
     monkeypatch.setattr(portfolio_utils, "_PRICE_SNAPSHOT", snapshot, raising=False)
-    monkeypatch.setattr("backend.common.instrument_api._resolve_full_ticker", lambda t, _: ("USDX", "US"))
-    monkeypatch.setattr(portfolio_utils, "get_instrument_meta", lambda _: {"currency": "USD", "name": "USD X"})
+    monkeypatch.setattr(
+        "backend.common.instrument_api._resolve_full_ticker", lambda t, _: ("USDX", "US")
+    )
+    monkeypatch.setattr(
+        portfolio_utils, "get_instrument_meta", lambda _: {"currency": "USD", "name": "USD X"}
+    )
     monkeypatch.setattr(portfolio_utils, "get_security_meta", lambda _: {})
     monkeypatch.setattr("backend.common.instrument_api.price_change_pct", lambda *_: None)
-    monkeypatch.setattr(portfolio_utils, "_fx_to_base", lambda c, b, cache: 0.8 if c == "USD" else 1.0)
+    monkeypatch.setattr(
+        portfolio_utils, "_fx_to_base", lambda c, b, cache: 0.8 if c == "USD" else 1.0
+    )
 
     portfolio = {
         "accounts": [
             {
                 "holdings": [
-                    {"ticker": ticker, "exchange": "US", "units": 2, "cost_basis_gbp": 100.0, "currency": "USD"}
+                    {
+                        "ticker": ticker,
+                        "exchange": "US",
+                        "units": 2,
+                        "cost_basis_gbp": 100.0,
+                        "currency": "USD",
+                    }
                 ]
             }
         ]
@@ -238,21 +362,31 @@ def test_last_close_fallback_snapshot_does_not_double_convert_fx(monkeypatch: py
     assert rows[0]["market_value_gbp"] == pytest.approx(160.0)
 
 
-def test_last_close_fallback_snapshot_marks_gbx_prices_as_gbp(monkeypatch: pytest.MonkeyPatch) -> None:
+def test_last_close_fallback_snapshot_marks_gbx_prices_as_gbp(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
     """GBX instruments should not be divided twice when snapshot uses last-close fallback."""
 
     ticker = "VOD.L"
     monkeypatch.setattr(prices, "_load_latest_prices", lambda _: {ticker: 1.103})
     monkeypatch.setattr(prices, "load_live_prices", lambda _: {})
     monkeypatch.setattr(prices, "_close_on", lambda *_: 1.103)
-    monkeypatch.setattr(prices.instrument_api, "_resolve_full_ticker", lambda full, latest: ("VOD", "L"))
+    monkeypatch.setattr(
+        prices.instrument_api, "_resolve_full_ticker", lambda full, latest: ("VOD", "L")
+    )
 
     snapshot = prices.get_price_snapshot([ticker])
     assert snapshot[ticker]["price_currency"] == "GBP"
 
     monkeypatch.setattr(portfolio_utils, "_PRICE_SNAPSHOT", snapshot, raising=False)
-    monkeypatch.setattr("backend.common.instrument_api._resolve_full_ticker", lambda t, _: ("VOD", "L"))
-    monkeypatch.setattr(portfolio_utils, "get_instrument_meta", lambda _: {"currency": "GBX", "name": "Vodafone"})
+    monkeypatch.setattr(
+        "backend.common.instrument_api._resolve_full_ticker", lambda t, _: ("VOD", "L")
+    )
+    monkeypatch.setattr(
+        portfolio_utils,
+        "get_instrument_meta",
+        lambda _: {"currency": "GBX", "name": "Vodafone"},
+    )
     monkeypatch.setattr(portfolio_utils, "get_security_meta", lambda _: {"currency": "GBX"})
     monkeypatch.setattr("backend.common.instrument_api.price_change_pct", lambda *_: None)
 
@@ -260,7 +394,13 @@ def test_last_close_fallback_snapshot_marks_gbx_prices_as_gbp(monkeypatch: pytes
         "accounts": [
             {
                 "holdings": [
-                    {"ticker": ticker, "exchange": "L", "units": 290, "cost_basis_gbp": 200.46, "currency": "GBX"}
+                    {
+                        "ticker": ticker,
+                        "exchange": "L",
+                        "units": 290,
+                        "cost_basis_gbp": 200.46,
+                        "currency": "GBX",
+                    }
                 ]
             }
         ]
@@ -284,6 +424,8 @@ def test_close_on_returns_none_when_fx_lookup_is_invalid(monkeypatch: pytest.Mon
     from backend.common import portfolio_utils
 
     monkeypatch.setattr(portfolio_utils, "_fx_to_base", lambda *_: None)
-    monkeypatch.setattr("backend.common.instruments.get_instrument_meta", lambda *_: {"currency": "USD"})
+    monkeypatch.setattr(
+        "backend.common.instruments.get_instrument_meta", lambda *_: {"currency": "USD"}
+    )
 
     assert prices._close_on("USDX", "US", sample_date) is None

--- a/tests/backend/common/test_prices.py
+++ b/tests/backend/common/test_prices.py
@@ -11,7 +11,7 @@ import pandas as pd
 import pytest
 
 from backend.common import portfolio_utils, prices
-from backend.common.portfolio_utils import PRICES_S3_KEY
+from backend.common.portfolio_utils import DATA_BUCKET_ENV, PRICES_S3_KEY
 
 
 def test_close_on_falls_back_to_close_column(monkeypatch: pytest.MonkeyPatch) -> None:
@@ -161,6 +161,15 @@ def test_get_price_snapshot_handles_stale_and_missing_data(monkeypatch: pytest.M
     assert info_ghi["change_7d_pct"] == pytest.approx((40.0 / 39.0 - 1.0) * 100.0)
     assert info_ghi["change_30d_pct"] == pytest.approx((40.0 / 38.0 - 1.0) * 100.0)
 
+    assert requested == [
+        ("ABC", "L", seven_day),
+        ("ABC", "L", thirty_day),
+        ("DEF", "N", seven_day),
+        ("DEF", "N", thirty_day),
+        ("GHI", "L", seven_day),
+        ("GHI", "L", thirty_day),
+    ]
+
 
 # ---------------------------------------------------------------------------
 # refresh_prices() S3 upload behaviour
@@ -183,7 +192,7 @@ def test_refresh_prices_uploads_to_s3_in_aws_env(
     """refresh_prices() must call s3.put_object with PRICES_S3_KEY when app_env==aws."""
     _stub_refresh_prices(tmp_path, monkeypatch)
     monkeypatch.setattr(prices.config, "app_env", "aws", raising=False)
-    monkeypatch.setenv("DATA_BUCKET", "test-bucket")
+    monkeypatch.setenv(DATA_BUCKET_ENV, "test-bucket")
 
     put_calls: list[dict] = []
 
@@ -211,7 +220,7 @@ def test_refresh_prices_s3_upload_failure_logs_warning_not_error(
     """If the S3 upload fails, refresh_prices() must log WARNING and not re-raise."""
     _stub_refresh_prices(tmp_path, monkeypatch)
     monkeypatch.setattr(prices.config, "app_env", "aws", raising=False)
-    monkeypatch.setenv("DATA_BUCKET", "test-bucket")
+    monkeypatch.setenv(DATA_BUCKET_ENV, "test-bucket")
 
     class FakeS3:
         def put_object(self, **kwargs):
@@ -239,7 +248,7 @@ def test_refresh_prices_skips_s3_when_data_bucket_not_set(
     """refresh_prices() must log a WARNING and skip S3 when DATA_BUCKET env var is absent."""
     _stub_refresh_prices(tmp_path, monkeypatch)
     monkeypatch.setattr(prices.config, "app_env", "aws", raising=False)
-    monkeypatch.delenv("DATA_BUCKET", raising=False)
+    monkeypatch.delenv(DATA_BUCKET_ENV, raising=False)
 
     import logging
 

--- a/tests/common/test_load_snapshot.py
+++ b/tests/common/test_load_snapshot.py
@@ -1,9 +1,7 @@
 import json
 import sys
-from datetime import datetime
 import types
-
-import pytest
+from datetime import datetime
 
 import backend.common as _backend_common
 from backend.common import portfolio_utils as pu
@@ -166,6 +164,48 @@ def test_load_snapshot_aws_s3_fails_no_local_path_configured_logs_error(
     assert "no local fallback configured" in caplog.text
 
 
+def _make_access_denied_s3_modules():
+    """Return (fake_boto3, fake_botocore_exceptions) that raise AccessDenied ClientError."""
+
+    class ClientError(Exception):
+        def __init__(self, msg):
+            super().__init__(msg)
+            self.response = {"Error": {"Code": "AccessDenied", "Message": msg}}
+
+    class FakeS3:
+        def get_object(self, Bucket, Key):  # noqa: N802
+            raise ClientError("Access Denied")
+
+    fake_boto3 = types.SimpleNamespace(client=lambda service: FakeS3())
+    fake_exc = types.SimpleNamespace(BotoCoreError=Exception, ClientError=ClientError)
+    return fake_boto3, fake_exc
+
+
+def test_load_snapshot_access_denied_logs_error_not_warning(tmp_path, monkeypatch, caplog):
+    """Non-NoSuchKey ClientError (e.g., AccessDenied) must still log at ERROR level."""
+    missing = tmp_path / "missing.json"
+    fake_boto3, fake_exc = _make_access_denied_s3_modules()
+
+    monkeypatch.setattr(pu.config, "app_env", "aws")
+    monkeypatch.setenv(pu.DATA_BUCKET_ENV, "bucket")
+    monkeypatch.setitem(sys.modules, "boto3", fake_boto3)
+    monkeypatch.setitem(sys.modules, "botocore.exceptions", fake_exc)
+    monkeypatch.setattr(pu.config, "prices_json", missing)
+    monkeypatch.setattr(pu, "_PRICES_PATH", missing)
+
+    import logging
+
+    with caplog.at_level(logging.WARNING):
+        data, ts = pu._load_snapshot()
+
+    assert data == {}
+    assert ts is None
+    error_records = [r for r in caplog.records if r.levelno >= logging.ERROR]
+    assert error_records, "AccessDenied must log at ERROR level, not WARNING"
+    assert "Failed to fetch price snapshot" in caplog.text
+    assert "not yet present" not in caplog.text
+
+
 def _make_no_such_key_s3_modules(tmp_path):
     """Return (fake_boto3, fake_botocore_exceptions) that raise NoSuchKey ClientError."""
 
@@ -208,7 +248,9 @@ def test_load_snapshot_nosuchkey_logs_warning_not_error(tmp_path, monkeypatch, c
     assert "Failed to fetch price snapshot" not in caplog.text
     assert "No price data available" not in caplog.text
     records = [r for r in caplog.records if r.levelno >= logging.ERROR]
-    assert records == [], f"Expected no ERROR logs for NoSuchKey; got: {[r.message for r in records]}"
+    assert records == [], (
+        f"Expected no ERROR logs for NoSuchKey; got: {[r.message for r in records]}"
+    )
 
 
 def test_load_snapshot_nosuchkey_no_local_path_logs_single_warning(

--- a/tests/common/test_load_snapshot.py
+++ b/tests/common/test_load_snapshot.py
@@ -166,6 +166,79 @@ def test_load_snapshot_aws_s3_fails_no_local_path_configured_logs_error(
     assert "no local fallback configured" in caplog.text
 
 
+def _make_no_such_key_s3_modules(tmp_path):
+    """Return (fake_boto3, fake_botocore_exceptions) that raise NoSuchKey ClientError."""
+
+    class ClientError(Exception):
+        def __init__(self, msg):
+            super().__init__(msg)
+            self.response = {"Error": {"Code": "NoSuchKey", "Message": msg}}
+
+    class FakeS3:
+        def get_object(self, Bucket, Key):  # noqa: N802
+            raise ClientError("The specified key does not exist.")
+
+    fake_boto3 = types.SimpleNamespace(client=lambda service: FakeS3())
+    fake_exc = types.SimpleNamespace(BotoCoreError=Exception, ClientError=ClientError)
+    return fake_boto3, fake_exc
+
+
+def test_load_snapshot_nosuchkey_logs_warning_not_error(tmp_path, monkeypatch, caplog):
+    """NoSuchKey S3 error must log WARNING (not ERROR) — expected on first deploy."""
+    missing = tmp_path / "missing.json"
+    fake_boto3, fake_exc = _make_no_such_key_s3_modules(tmp_path)
+
+    monkeypatch.setattr(pu.config, "app_env", "aws")
+    monkeypatch.setenv(pu.DATA_BUCKET_ENV, "bucket")
+    monkeypatch.setitem(sys.modules, "boto3", fake_boto3)
+    monkeypatch.setitem(sys.modules, "botocore.exceptions", fake_exc)
+    monkeypatch.setattr(pu.config, "prices_json", missing)
+    monkeypatch.setattr(pu, "_PRICES_PATH", missing)
+
+    import logging
+
+    with caplog.at_level(logging.WARNING):
+        data, ts = pu._load_snapshot()
+
+    assert data == {}
+    assert ts is None
+    # Only one WARNING — no ERROR log
+    assert "not yet present in S3" in caplog.text
+    assert "not yet seeded" in caplog.text
+    assert "Failed to fetch price snapshot" not in caplog.text
+    assert "No price data available" not in caplog.text
+    records = [r for r in caplog.records if r.levelno >= logging.ERROR]
+    assert records == [], f"Expected no ERROR logs for NoSuchKey; got: {[r.message for r in records]}"
+
+
+def test_load_snapshot_nosuchkey_no_local_path_logs_single_warning(
+    monkeypatch, caplog
+):
+    """NoSuchKey + no local path configured: single WARNING, no ERROR."""
+    fake_boto3, fake_exc = _make_no_such_key_s3_modules(None)
+
+    monkeypatch.setattr(pu.config, "app_env", "aws")
+    monkeypatch.setenv(pu.DATA_BUCKET_ENV, "bucket")
+    monkeypatch.setitem(sys.modules, "boto3", fake_boto3)
+    monkeypatch.setitem(sys.modules, "botocore.exceptions", fake_exc)
+    monkeypatch.setattr(pu.config, "prices_json", None)
+    monkeypatch.setattr(pu, "_PRICES_PATH", None)
+
+    import logging
+
+    with caplog.at_level(logging.WARNING):
+        data, ts = pu._load_snapshot()
+
+    assert data == {}
+    assert ts is None
+    warning_texts = [r.message for r in caplog.records if r.levelno == logging.WARNING]
+    error_texts = [r.message for r in caplog.records if r.levelno >= logging.ERROR]
+    assert error_texts == [], f"Expected no ERROR logs for NoSuchKey; got: {error_texts}"
+    assert any("not yet" in m for m in warning_texts), (
+        f"Expected a 'not yet seeded' warning; got: {warning_texts}"
+    )
+
+
 def test_load_snapshot_non_aws_missing_local_logs_only_warning(
     tmp_path, monkeypatch, caplog
 ):

--- a/tests/common/test_load_snapshot.py
+++ b/tests/common/test_load_snapshot.py
@@ -206,7 +206,7 @@ def test_load_snapshot_access_denied_logs_error_not_warning(tmp_path, monkeypatc
     assert "not yet present" not in caplog.text
 
 
-def _make_no_such_key_s3_modules(tmp_path):
+def _make_no_such_key_s3_modules():
     """Return (fake_boto3, fake_botocore_exceptions) that raise NoSuchKey ClientError."""
 
     class ClientError(Exception):
@@ -226,7 +226,7 @@ def _make_no_such_key_s3_modules(tmp_path):
 def test_load_snapshot_nosuchkey_logs_warning_not_error(tmp_path, monkeypatch, caplog):
     """NoSuchKey S3 error must log WARNING (not ERROR) — expected on first deploy."""
     missing = tmp_path / "missing.json"
-    fake_boto3, fake_exc = _make_no_such_key_s3_modules(tmp_path)
+    fake_boto3, fake_exc = _make_no_such_key_s3_modules()
 
     monkeypatch.setattr(pu.config, "app_env", "aws")
     monkeypatch.setenv(pu.DATA_BUCKET_ENV, "bucket")
@@ -257,7 +257,7 @@ def test_load_snapshot_nosuchkey_no_local_path_logs_single_warning(
     monkeypatch, caplog
 ):
     """NoSuchKey + no local path configured: single WARNING, no ERROR."""
-    fake_boto3, fake_exc = _make_no_such_key_s3_modules(None)
+    fake_boto3, fake_exc = _make_no_such_key_s3_modules()
 
     monkeypatch.setattr(pu.config, "app_env", "aws")
     monkeypatch.setenv(pu.DATA_BUCKET_ENV, "bucket")
@@ -279,6 +279,41 @@ def test_load_snapshot_nosuchkey_no_local_path_logs_single_warning(
     assert any("not yet" in m for m in warning_texts), (
         f"Expected a 'not yet seeded' warning; got: {warning_texts}"
     )
+
+
+def test_load_snapshot_botocore_error_logs_error(tmp_path, monkeypatch, caplog):
+    """BotoCoreError (non-ClientError) must still log at ERROR level after the refactor."""
+    missing = tmp_path / "missing.json"
+
+    class BotoCoreError(Exception):
+        pass
+
+    class FakeS3:
+        def get_object(self, Bucket, Key):  # noqa: N802
+            raise BotoCoreError("connection timeout")
+
+    fake_boto3 = types.SimpleNamespace(client=lambda service: FakeS3())
+    # BotoCoreError is used as both BotoCoreError and ClientError base in the except clause
+    fake_exc = types.SimpleNamespace(BotoCoreError=BotoCoreError, ClientError=Exception)
+
+    monkeypatch.setattr(pu.config, "app_env", "aws")
+    monkeypatch.setenv(pu.DATA_BUCKET_ENV, "bucket")
+    monkeypatch.setitem(sys.modules, "boto3", fake_boto3)
+    monkeypatch.setitem(sys.modules, "botocore.exceptions", fake_exc)
+    monkeypatch.setattr(pu.config, "prices_json", missing)
+    monkeypatch.setattr(pu, "_PRICES_PATH", missing)
+
+    import logging
+
+    with caplog.at_level(logging.WARNING):
+        data, ts = pu._load_snapshot()
+
+    assert data == {}
+    assert ts is None
+    error_records = [r for r in caplog.records if r.levelno >= logging.ERROR]
+    assert error_records, "BotoCoreError must log at ERROR level"
+    assert "Failed to fetch price snapshot" in caplog.text
+    assert "not yet present" not in caplog.text
 
 
 def test_load_snapshot_non_aws_missing_local_logs_only_warning(

